src/services/plan-service.ts

```diff
@@ -0,0 +1,257 @@
+  planList: Plan[],
+  { minFee, minSubscriberLimit }: { minFee: number; minSubscriberLimit: number } = {
+    minFee: 0,
+    minSubscriberLimit: 0,
+  },
+): SubscribersLimitedPlan[] =>
+  planList.filter(
+    (x) => x.type === 'subscribers' && x.fee > minFee && x.subscriberLimit >= minSubscriberLimit,
+  ) as SubscribersLimitedPlan[];
+
+const getFreePlans = (planList: Plan[]): FreePlan[] =>
+  planList.filter((x) => x.type === 'free') as FreePlan[];
+
+const getAgencyPlans = (planList: Plan[]): AgencyPlan[] =>
+  planList.filter((x) => x.type === 'agency') as AgencyPlan[];
+
+const getStandardPlans = (
+  planList: Plan[],
+): (MonthlyRenewalDeliveriesPlan | SubscribersLimitedPlan | PrepaidPack)[] =>
+  planList.filter((x) => x.featureSet === 'standard') as (
+    | MonthlyRenewalDeliveriesPlan
+    | SubscribersLimitedPlan
+    | PrepaidPack
+  )[];
+
+const getPlusPlans = (
+  planList: Plan[],
+): (MonthlyRenewalDeliveriesPlan | SubscribersLimitedPlan)[] =>
+  planList.filter((x) => x.featureSet === 'plus') as (
+    | MonthlyRenewalDeliveriesPlan
+    | SubscribersLimitedPlan
+  )[];
+
+const getFreePathOrEmpty = (userPlan: Plan, planList: Plan[]): [] | [FreePath] => {
+  var plans = getFreePlans(planList);
+
+  if (plans.length === 0) {
+    return [];
+  }
+
+  var cheapestPlan = plans[0];
+
+  return [
+    {
+      type: 'free',
+      actual: userPlan.featureSet === 'free',
+      deadEnd: plans.length === 1 && userPlan === cheapestPlan,
+    },
+  ];
+};
+
+const getAgencyPathOrEmpty = (userPlan: Plan, planList: Plan[]): [] | [AgenciesPath] => {
+  var plans = getAgencyPlans(planList);
+
+  if (plans.length === 0) {
+    return [];
+  }
+
+  var cheapestPlan = plans[0];
+
+  return [
+    {
+      type: 'agencies',
+      actual: userPlan.featureSet === 'agency',
+      deadEnd: plans.length === 1 && userPlan === cheapestPlan,
+    },
+  ];
+};
+
+const getStandardPathOrEmpty = (userPlan: Plan, planList: Plan[]): [] | [StandardPath] => {
+  var plans = getStandardPlans(planList).sort((x) => getPlanFee(x));
+
+  if (plans.length === 0) {
+    return [];
+  }
+
+  var cheapestPlan = plans[0];
+
+  return [
+    {
+      type: 'standard',
+      actual: userPlan.featureSet === 'standard',
+      minimumFee: getPlanFee(cheapestPlan),
+      deadEnd: plans.length === 1 && cheapestPlan === userPlan,
+    },
+  ];
+};
+
+const getPlusPathOrEmpty = (userPlan: Plan, planList: Plan[]): PlusPath[] => {
+  var plans = getPlusPlans(planList).sort((x) => x.fee);
+
+  if (plans.length === 0) {
+    return [];
+  }
+
+  var cheapestPlan = plans[0];
+
+  return [
+    {
+      type: 'plus',
+      actual: userPlan.featureSet === 'plus',
+      minimumFee: cheapestPlan.fee,
+      deadEnd: plans.length === 1 && cheapestPlan === userPlan,
+    },
+  ];
+};
+
+const _agencyPlan: AgencyPlan = {
+  type: 'agency',
+  featureSet: 'agency',
+};
+
+const getPotentialUpgrades = (userPlan: Plan, planList: Plan[]): Plan[] => {
```

Comment:
Is it ok to use the word `potential`? 🤔 I think about 'potencial' as something that we don't have yet, something that could to be in a future. Maybe I don't have the concept, but instead potential, what about the words 'possible' or 'available'

Reply:
_Available_ sounds good!

Reply:
https://www.merriam-webster.com/dictionary/potential
According to your definition and this one, it is a possible thing, and a future plan is a possible thing. I don´t get your point 

potential adjective
1: existing in possibility : capable of development into actuality

possible adjective
...
3: having an indicated potential
🤔

Reply:
https://www.merriam-webster.com/dictionary/available 🤔 
I still think potential is more suitable for this case. You can try and convince me though 😛

Reply:
Yes, maybe 'potential' sounds me a little weird. It seems to be that the use here is ok but I think about 'potential' as something that is not defined yet.